### ISSUE TITLE
URL escaping wrongly picks up > character

### ISSUE THREAD

**Issue by epa** [rjbs/HTML-FromText#2]
Sometimes <> are used to delimit URIs in plain text documents.  This works well since neither is a valid character in a URI.  However, HTML::FromText wrongly picks up the > as part of the URI:

````
my $t2h = new HTML::FromText({ urls => 1 });
my $text = 'hello <http://whatever.com/xyz> there';
say $t2h->parse($text);
````

This produces the rather garbled output

````
&lt;<a href="http://whatever.com/xyz&gt" class="hft-urls">http://whatever.com/xyz&gt</a>;
````

There are two bugs here: one is that > is being picked up as part of the URL, when it cannot be.  The URL-sniffing should restrict itself to characters that can appear in a URL.

An additional bug is the strange escaping, where `</a>` has been sandwiched inside `&gt;` somehow.  The earlier `&gt;` is also missing its semicolon.

**Comment by rjbs** [comment]
Ha, nice!  I'm not sure when/whether I'll get to this, as mostly I'm here to shepherd patches.  Sounds like it might be fun to have a crack at, though.

**Comment by epa** [comment]
I suggest tweaking the regexp in routine `urls`.  Instead of `[w+/]` it should have the legal set of characters in a URI.  Then the trailing > would not get picked up and the escaping bug wouldn't get tickled either.

**Comment by rjbs** [comment]
You're mistaken about the cause, I'm afraid, but I _did_ push a commit to master to only allow URI-legal characters in URIs.

The problem is that by the time that `urls` is being called, `<http://foo>` has already been rewritten to `&lt;http://foo&gt;`, and `http://foo&gt;` is a valid URI, composed of legal characters.

Presumably the timing of `metachars` is wrong, but I think that fixing it correctly will require some untangling.  I'm stopping work for now, though.

**Comment by epa** [comment]
Blarg.  If the &-escaping happens before looking for URIs, then we know that only `&amp;` can appear in URIs.  Other things such as `&lt;` or `&foo` cannot appear.  Perhaps the URI regexp can be changed to something like

```
([a-zA-Z0-9,....]|&amp;)+
```

to allow all URI-legal characters except `&`, and allow `&amp;`.
